File: php/disassembler.php
<?
/* This program converts Hack binary into Hack Assembly Language.
To run, type: php disassembler.php <fileName.hack>

This program is intended for use in PHP CLI*/

function output($message)  /*function to output to console*/
{
    $message = date("H:i:s") . " - $message ".PHP_EOL;
    print($message);
    flush();
}

if ($argc < 2){ // check if any arguments provided, exit if not
  output("ERROR: No input file provided. Use php disassembler.php <input file>");
  die("");
}

if (strpos($argv[1],".hack")==false){ // check if valid file type provided as argument
  output("ERROR: Input file must be a .hack file");
  die("");
}

$compTable = [ // create lookup table
  '101010' => '0',
  '111111' => '1',
  '111010' => '-1',
  '001100' => 'D',
  '110000' => 'A,M',
  '001101' => '!D',
  '110001' => '!A,!M',
  '001111' => '-D',
  '110011' => '-A,-M',
  '011111' => 'D+1',
  '110111' => 'A+1,M+1',
  '001110' => 'D-1',
  '110010' => 'A-1,M-1',
  '000010' => 'D+A,D+M',
  '010011' => 'D-A,D-M',
  '000111' => 'A-D,M-D',
  '000000' => 'D&A,D&M',
  '010101' => 'D|A,D|M'
];

$destTable = [ // create lookup table
  '000' => '',
  '001' => 'M=',
  '010' => 'D=',
  '011' => 'DM=',
  '100' => 'A=',
  '101' => 'AM=',
  '110' => 'AD=',
  '111' => 'ADM='
];

$jumpTable = [ // create lookup table
  '000' => '',
  '001' => ';JGT',
  '010' => ';JEQ',
  '011' => ';JGE',
  '100' => ';JLT',
  '101' => ';JNE',
  '110' => ';JLE',
  '111' => ';JMP'
];

$rawData = file($argv[1]); //open/file
$outputArray = []; //define array to store translated instructions
$i = 1; //counter for data processing
foreach($rawData as $line){ // iterate/operate on each line
  if ($line[0] == '0'){ // if 0, then A instruction
    $val = substr($line,1,15); // splice line to get destination
    $val = bindec($val); // convert to binary
    $val = strval($val); // convert back to string
    $instruction = '@' . $val . "\n"; // build instruction
    $outputArray[]=$instruction; // add instruction to array
  }
  else if($line[0] == '1'){ // else if 1, then C instruction
    $aBit = $line[3];
    $compBits = substr($line, 4, 6); // Extracts characters from index 4 to 9
    $destBits = substr($line, 10, 3); // Extracts characters from index 10 to 12
    $jmpBits = substr($line, 13, 3); // Extracts characters from index 13 to 15
    $dest = $destTable[$destBits]; // convert destbits
    $jmp = $jumpTable[$jmpBits]; // convert jumpbits
    $compUnfiltered = $compTable[$compBits]; // get compbits
    $compUnfilteredParts = explode(',', $compUnfiltered); // explode compbits
    $aBitIndex = intval($aBit); // convert to int
    $comp = $compUnfilteredParts[$aBitIndex]; // find specific comp value
    $instruction = $dest . $comp . $jmp . "\n"; // build instruction
    $outputArray[]=$instruction; // add instruction to array
  }
  else{ // if no dest bit and unable to detect A or C function
    output("ERROR: Unable to parse line $i");
    die("");
  }
  $i = $i + 1; // increment line counter
}
// Replace the extension .hack with .asm in the file name
$filename = str_replace('.hack', '.asm', $argv[1]);

// Open the file in write mode
$outfile = fopen($filename, 'w');
// Write the lines from $outputArray to the file
foreach ($outputArray as $line) {
    fwrite($outfile, $line);
}
fclose($outfile); // close file
output("File written to $filename"); // output success message
?>

File: Python/disassembler.py
#  This program converts Hack binary into Hack Assembly Language.
# To run, type: python disassembler.py <fileName.hack>
# if on a Linux system, you may need to run it with: python3 disassembler.py <fileName.hack>

import sys
from os.path import exists as file_exists

if len(sys.argv) < 2:  # check that argument/input file provided
  print(
      "ERROR: No input file provided. Use python disassembler.py <fileName.hack>"
  )
  exit(1)
# Check if appropriate file extension
if ".hack" not in sys.argv[1]:
  print("ERROR: Input file must be a Hack file.")
  exit(1)
else:
  if file_exists(sys.argv[1]): # if file exists, then read lines from it
    file = open(sys.argv[1], 'r')
    Lines = file.readlines()

    # List to be used for storing HACK instructions
    hackList = []

    # Computation Dictionary
    compTable = {
        '101010': '0',
        '111111': '1',
        '111010': '-1',
        '001100': 'D',
        '110000': 'A,M',
        '001101': '!D',
        '110001': '!A,!M',
        '001111': '-D',
        '110011': '-A,-M',
        '011111': 'D+1',
        '110111': 'A+1,M+1',
        '001110': 'D-1',
        '110010': 'A-1,M-1',
        '000010': 'D+A,D+M',
        '010011': 'D-A,D-M',
        '000111': 'A-D,M-D',
        '000000': 'D&A,D&M',
        '010101': 'D|A,D|M'
    }

    # Destination Dictionary
    destTable = {
        '000': '',
        '001': 'M=',
        '010': 'D=',
        '011': 'DM=',
        '100': 'A=',
        '101': 'AM=',
        '110': 'AD=',
        '111': 'ADM='
    }

    # Jump Dictionary
    jumpTable = {
        '000': '',
        '001': ';JGT',
        '010': ';JEQ',
        '011': ';JGE',
        '100': ';JLT',
        '101': ';JNE',
        '110': ';JLE',
        '111': ';JMP'
    }

    # Loop through all assembly lines in the HACK file
    for line in Lines:
      if line[0] == '0': # if first bit is 0, then it is an A instruction
        val = str(int(line[1:16], 2))  # Get rest of string, convert to decimal, then to string
        instruction = '@' + val + '\n'  # build instruction
        hackList.append(instruction) # add instruction to list

      elif line[0] == '1': # if first bit is 1, then this is a C instruction
        aBit = line[3] # splice for retrieving the A bits
        compBits = line[4:10] # splice to get compBits
        destBits = line[10:13] #splice to get destbits
        jmpBits = line[13:16] # splice out jumpbits
        dest = destTable[destBits] # get destination string from destTable
        compUnfiltered = compTable[compBits] #grab unfiltered computation from compTable
        comp = compUnfiltered.split(',')[int(aBit)] # use aBit to determine the comp instruction
        jmp = jumpTable[jmpBits] # grab jump location from jumpTable
        instruction = dest + comp + jmp + '\n' # build instruction
        hackList.append(instruction) # add instruction to list
        
    file = open(sys.argv[1].replace('.hack', '.asm'), 'w') # create and open file
    file.writelines(hackList) # write to file
    print("File written to " + sys.argv[1].replace('.hack', '.asm'))
    file.close()

  else: # if file does not exist, then print error
    print("ERROR: File " + sys.argv[1] + " does not exist.")
    exit(1)
